queryable-spec.ts

```diff
@@ -0,0 +1,427 @@
+ * (bindings orderings) interfaces. Furthermore, QueryOperationObjects are
+ * used to represent both available orderings (i.e. orderings that may be
+ * requested by callers) and returned orderings (i.e. orderings followed
+ * by the returned iterators).
+ */
+interface QueryOperationOrder<T extends TermName | RDF.Variable> {
+  cost: QueryOperationCost;
+  terms: { term: T, direction: 'asc' | 'desc' }[];
+}
+
+interface QueryResultCardinality {
+  /**
+   * indicates the type of counting that was done, and MUST either be 
+   * "estimate" or "exact".
+   */
+  type: 'estimate' | 'exact';
+
+  /**
+   * Indicates an estimate of the number of quads in the stream if 
+   * type = "estimate", or the exact number of quads in the stream if 
+   * type = "exact".
+   */
+  value: number;
+}
+
+/**
+ * A QueryResultMetadata is an object that contains metadata about a certain
+ * query result.
+ */
+interface QueryResultMetadata<OrderItemsType extends TermName | RDF.Variable> {
+
+  /**
+   * An optional field that contains metadata about the number of quads in the
+   * result stream.
+   */
+  cardinality?: QueryResultCardinality;
+
+  /**
+   * An optional field that contains the available options for quad sorting
+   * based on the provided pattern, expression and options.
+   */
+  availableOrders?: QueryOperationOrder<OrderItemsType>[];
+
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+}
+
+/**
+ * A QueryResultMetadataOptions is an object that gives suggestions on what
+ * type of metadata is desired, such as when invoking FilterResult.metadata.
+ */
+ interface QueryResultMetadataOptions {
+
+  /**
+   * optional field that MAY either contain "estimate" or "exact". If defined,
+   * this type MUST correspond to the type in QueryResultMetadataCardinality.
+   */
+  cardinality?: 'estimate' | 'exact';
+
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+};
+
+/**
+ * Generic interface that defines the API pattern for query result objects.
+ */
+interface BaseQueryResult {
+  
+  type: string;
+
+  /**
+   * Returns either a stream containing all the items that match the given query,
+   * a boolean or void depending on the semantics of the given query.
+   */
+  execute(opts?: any): Promise<Stream<any> | boolean | void>;
+
+  /**
+   * Asynchronously metadata of the current result.
+   */
+  metadata?(opts?: QueryResultMetadataOptions): Promise<QueryResultMetadata<any>>;
+
+  /**
+   * Asynchronously returns a boolean indicating if the requested expression is
+   * supported. If it returns true, execute() and metadata() MAY produce valid 
+   * results. If it returns false, execute() MUST return a stream emitting an 
+   * error, and metadata() MUST reject.
+   */
+  isSupported(): Promise<boolean>;
+}
+
+
+interface QueryResultBindings extends BaseQueryResult {
+  type: 'bindings';
+  execute(opts?: { order?: QueryOperationOrder<RDF.Variable> }): Promise<Stream<Bindings>>;
+  variables: RDF.Variable[];
```

Comment:
@rubensworks if we could move `variables` into the return value of `metadata()` we'd have a completely uniform interface across all types of results. Any objection?

Reply:
Hm, good idea.
We may want to make it a mandatory metadata entry for the bindings then though.